python/gsw/main.py

```diff
@@ -0,0 +1,370 @@
+        return json.loads(data)
+
+    def check_for_email(self):
+        '''
+        Reads the most recent unread email from the Iridium email account. 
+
+        If it is a downlink, the function returns a statefield report. If it is an uplink 
+        confirmation, the function records that we recieved an uplink confirmation and returns
+        None. If there is neither a downlink or an uplink, the function returns None.
+
+        If either a downlink or an uplink is recieved, the function will update the most recent MOMSN 
+        and MTMSN numbers and check whether or not radioSession can send uplinks.
+        '''
+        #look for all new emails from iridium
+        _, data = self.mail.search(None, '(FROM "sbdservice@sbd.iridium.com")', '(UNSEEN)')
+        mail_ids = data[0]
+        id_list = mail_ids.split()
+
+        for num in id_list:
+            #.fetch() fetches the mail for given id where 'RFC822' is an Internet 
+            # Message Access Protocol.
+            _, data = self.mail.fetch(num,'(RFC822)')
+
+            #go through each component of data
+            for response_part in data:
+                if isinstance(response_part, tuple):
+                    # converts message from byte literal to string removing b''
+                    msg = email.message_from_string(response_part[1].decode('utf-8'))
+                    email_subject = msg['subject']
+
+                    #handles uplink confirmations
+                    if email_subject.find("SBD Mobile Terminated Message Queued for Unit: ")==0:
+                        # Get imei number of the radio that recieved the uplink
+                        self.imei=int(email_subject[47:])
+
+                        for part in msg.walk():
+                                            
+                            if part.get_content_maintype() == 'multipart':
+                                continue
+
+                            if part.get('Content-Disposition') is None:
+                                continue
+
+                            #get the body text of the email and look for the MOMSN/MTMSN number
+                            if part.get_content_type() == "text/plain":
+                                email_body = part.get_payload(decode=True).decode('utf8')
+                                for line in email_body.splitlines():
+                                    if line.find("MTMSN")!=-1:
+                                        self.mtmsn=int(line[line.find("MTMSN")+9:line.find("MTMSN")+11])
+
+                            #Verify whether or not RadioSession can send uplinks
+                            if self.confirmation_mtmsn != self.mtmsn:
+                                #stop radio session from sending any more uplinks
+                                self.send_uplinks=False
+                            else:
+                                #allow radio session to send more uplinks
+                                self.send_uplinks=True
```

Comment:
I don't think we should do this here. Here's the problem I see: if `confirmation_mtmsn == self.mtmsn`, then it means that Iridium has queued up the most recent uplink, but it doesn't mean that the _satellite_ has received the uplink. Setting `send_uplinks` to true will therefore allow uplink packets to stack up on the Iridium servers, since RadioSession will just see a go-ahead to send more packets.

Maybe my logic is wrong here, though. Let me know if this sounds wrong

Reply:
So this is my understanding of how the Iridium email system works (I could be wrong, though): 

The uplink confirmation email gives the mtmsn of the message queued to be be sent to the satellite by Iridium. If the function identifies an uplink confirmation email, it will set the mtmsn class variable to the mtmsn of the most recent message queued for uplink.

The downlink email gives the mtmsn of most recent message that the satellite recieved. If the function identifies a downlink, it sets the confirmation_mtmsn variable to the mtmsn of the most recent message that the satellite recieved.

If the mtmsn and the confirmation_mtmsn match up, then radio session is cleared to send uplinks. I think that, by this system, we can be sure that we are only queuing one message to be sent to the satellite at a time.

Reply:
This makes sense! Sorry, I got confused between the meanings of `confirmation_mtmsn` and `mtmsn`.

Can the code that checks these variables' equality be moved below the email processing code? Otherwise it's duplicated across the uplink and downlink.

Reply:
I can't check the variables equality after I return the statefield report. Since this function only looks at the most recent email, we will be checking equality only once each time this function is called. So I don't think having it duplicated should be much of a problem.